beluga_tutorial/src/main.cpp

```diff
@@ -0,0 +1,258 @@
+
+  // Execute the particle filter using beluga
+  double current_pose{tutorial_params.initial_pose};
+  for (auto n = 0; n < tutorial_params.number_of_cycles; n++) {
+    // Prepare the data to be saved
+    TutorialData tutorial_data;
+
+    // Save ground truth
+    current_pose += (tutorial_params.velocity * tutorial_params.dt);
+    tutorial_data.ground_truth = current_pose;
+
+    // Check if the simulation is out of bounds
+    if (current_pose > tutorial_params.map_size) {
+      break;
+    }
+
+    // Motion model
+    auto motion_model = [&](double state) {
+      double distance = (tutorial_params.velocity * tutorial_params.dt);
+      double translation_param = generateRandom(0.0, tutorial_params.translation_sigma);
+      return state + distance - translation_param;
+    };
+
+    // Generate simulated sensor data
+    auto sensor_data = landmark_map | ranges::views::remove_if([&](const double lm) {
+                         return std::abs(lm - current_pose) > tutorial_params.sensor_range;
+                       }) |
+                       ranges::views::transform([&](const double lm) { return lm - current_pose; }) |
+                       ranges::to<SensorData>;
+    // Sensor model
+    auto sensor_model = [&](const double& state) {
+      auto particle_sensor_data =
+          landmark_map | ranges::views::transform([&](const double lm) { return lm - state; }) | ranges::to<SensorData>;
+
+      LandmarkMapVector landmark_probability;
+      for (const auto& sensor_datum : sensor_data) {
+        auto distances = particle_sensor_data | ranges::views::transform([&](auto& particle_sensor_datum) {
+                           return std::abs(sensor_datum - particle_sensor_datum);
+                         });
+
+        auto min_distance = ranges::min(distances);
+        landmark_probability.push_back(exp((-1 * pow(min_distance, 2)) / (2 * tutorial_params.sensor_model_sigam)));
+      }
+
+      return ranges::accumulate(landmark_probability, 1.0, std::multiplies<>{}) + tutorial_params.min_particle_weight;
+    };
+
+    // For the propose of the tutorial, we split the process in the following stages:
+    // Current stage
+    tutorial_data.current = particles;
+
+    // Propagate stage
+    particles |= beluga::actions::propagate(std::execution::seq, motion_model);
+    tutorial_data.propagate = particles;
+
+    // Reweight stage
+    particles |= beluga::actions::reweight(std::execution::seq, sensor_model) | beluga::actions::normalize;
+    tutorial_data.reweight = particles;
+
+    // Resample
+    particles |= beluga::views::sample | ranges::views::take_exactly(tutorial_params.number_of_particles) |
+                 beluga::actions::assign;
+    tutorial_data.resample = particles;
+
+    // Calculate mean and standard deviation
+    const auto estimation = beluga::estimate(beluga::views::states(particles), beluga::views::weights(particles));
+    tutorial_data.estimation = estimation;
```

Comment:
@Alondruck nit: can we collapse this two in a single line?

Reply:
I agree with you. I chose to keep it in two separate steps for the tutorial explanation. Since 'tutorial_data' is used for the visualization part, I wanted to simplify the tutorial by leaving that for another tutorial or for another more advanced step of the same tutorial.